---
title: "Project 5: Statistical phylogenetics"
date:  "`r format(Sys.time(), '%B %d, %Y')`"
author: "Team C - Minghang Li, Xiaocheng Yang, Xinyi Chen"
mainfont: "KpRoman"
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
output:
  pdf_document:
    latex_engine: xelatex
    pandoc_args: "--pdf-engine-opt=-shell-escape"
    extra_dependencies: ["amsmath", "amsthm", "amssymb", "cancel", "unicode-math" ]
---

# Problem 12: Transition matrix, rate matrix, and stationary distribution

## 1. Show that $\frac{dP(t)}{dt} = R \cdot P(t)$

```{=tex}
\begin{proof}

$$
\begin{aligned}
  \frac{dP(t)}{dt} &= \frac{P(dt + t) - P(t)}{dt} & (\text{Definition of derivative})\\
                   &= \frac{P(dt)P(t) - P(t)}{dt} & (\text{Chapman-Kolmogorov's equation}) \\
                   &= \frac{(I + Rdt)P(t) - P(t)}{dt} & (\text{Definition of }P(dt))\\
                   &= \frac{\cancel{P(t)} + RP(t)dt - \cancel{P(t)}}{dt} \\
                   &= \frac{RP(t)dt}{dt} \\
                   &= RP(t)
\end{aligned}
$$
\end{proof}
```
## 2. Assume that the given Markov chain is ergodic with (unique) stationary distribution $\pi$, show that $R\pi = 0$

```{=tex}
\begin{proof}
The stationary distribution $\pi$ satisfies:

$$
P(t)\pi = \pi
$$
after a sufficiently long time $t$ and any time point that follows. Hence, the
 following equation also holds:

$$
P(t + dt) \pi = \pi
$$

We can further re-write the left-hand side of the equation:

$$
\begin{aligned}
P(dt + t) \pi &= P(dt)P(t)\pi & (\text{Chapman-Kologomorov's equation})\\
              &= (I + Rdt)P(t)\pi & (\text{Definition of }P(dt)) \\
              &= P(t)\pi + R \cdot P(t) dt \pi \\
              &= \pi + Rdt \pi & (\text{Plug in }P(t)\pi = \pi)
\end{aligned}
$$

which leads us to the following equality:

$$
\begin{aligned}
\cancel{\pi} + Rdt\pi &= \cancel{\pi} \\
               Rdt\pi &= 0 \\
                R \pi &= 0
\end{aligned}
$$
\end{proof}
```
# Problem 13: Phylogenetic trees as Bayesian networks

## 1. What is the joint probability $P(X, Z|T)$ of the tree?

$$
P(X, Z|T) =     \pi(Z_4)
      P(X_5|Z_4)        P(Z_3|Z_4)
             P(Z_2|Z_3)            P(Z_1|Z_3)
        P(X_4|Z_2) P(X_3|Z_2) P(X_2|Z_1) P(X_1|Z_1)
$$

## 2. How many summation steps would be required for the naive calculation of $P(X|T)$ via brute-force marginalization over the hidden nodes $Z$?

Since there are 4 possibilities (A, C, G, T) for each hidden node:

$$
4 \times 4 \times 4 \times 4 = 256
$$

## 3. Rearrange the expression $P(X|T)$ such that the number of operations is minimized. How many summation steps are required now for the calculation of $P(X|T)$?

$$
\begin{aligned}
P(X|T) &= \sum_{Z_4}\sum_{Z_3}\sum_{Z_2}\sum_{Z_1}\pi(Z_4)P(X_5|Z_4)P(Z_3|Z_4)
P(Z_2|Z_3)P(Z_1|Z_3)P(X_4|Z_2) P(X_3|Z_2) P(X_2|Z_1) P(X_1|Z_1) \\
&= \sum_{Z_4} \pi(Z_4) P(X_5|Z_4)
\sum_{Z_3} P(Z_3|Z_4)
\sum_{Z_2} P(Z_2|Z_3) P(X_4|Z_2) P(X_3|Z_2)
\sum_{Z_1} P(Z_1|Z_3)P(X_2|Z_1) P(X_1|Z_1) & (\text{distributivity})\\
&= \underbrace{\sum_{Z_4} \pi(Z_4) P(X_5|Z_4)}_4
\underbrace{\sum_{Z_3} P(Z_3|Z_4)}_{4 \times 4}
\underbrace{\sum_{Z_2} P(Z_2|Z_3) P(X_4|Z_2) P(X_3|Z_2)}_{
  4 \times 4 (Z_2 \rightarrow X_4 \text{ branch})
  + 4 \times 4 (Z_2 \rightarrow X_3 \text{ branch})
}
\underbrace{\sum_{Z_1} P(Z_1|Z_3)P(X_2|Z_1) P(X_1|Z_1)}_{
  4 \times 4 (Z_1 \rightarrow X_2 \text{ branch})
  + 4 \times 4 (Z_1 \rightarrow X_1 \text{ branch})
} \\
\end{aligned}
$$

In total we have $(16+16) \times 2 + 16 + 4 = 52$ summations.

# Problem 14: Learning phylogenetic trees from sequence alignment data

```{r load_library, include=FALSE}
library(ape)
library(phangorn)
```

Load the dataset `ParisRT.txt`.

```{r read_data}
data <- read.dna("ParisRT.txt", format="sequential")
data
```

Create initial tree topology for the alignment using neighbor joining
and K80 model.

```{r create_ini_tree}
set.seed(42)
tree.ini <- NJ(dist.dna(data, model = "K80"))
```

Plot the initial tree.

```{r plot_initial_tree, fig.width=10, fig.height=5}
plot.phylo(tree.ini)
```

```{r fit_k80_model}
tree.ML <- pml(tree.ini, phyDat(data), model = "K80")
tree.ML
```

Find the optimal parameters of K80 model for rate matrix.

```{r optim_k80}
tree.optim.Q <- optim.pml(
  tree.ML,
  optQ = TRUE,
  optNni = FALSE,
  optBf = FALSE,
  optInv = FALSE,
  optGamma = FALSE,
  optEdge = FALSE,
  optRate = FALSE,
  optRooted = FALSE,
  model = "K80"
)
```

```{r print_optim_k80}
tree.optim.Q
```

Optimize for branch lengths, nucelotide substitution rates and tree
topology simultaneously.

```{r optim_several}
tree.optim <- optim.pml(
  tree.ML,
  optQ = TRUE,       # rate matrix
  optNni = TRUE,     # tree topology
  optBf = FALSE,     # base frequencies
  optInv = FALSE,    # proportion of var size
  optGamma = FALSE,  # gamma rate param
  optEdge = TRUE,    # edge lengths
  optRate = FALSE,   # overall rate
  optRooted = FALSE, # edge lengths of a rooted tree
  model = "K80"
)
```

```{r print_optim_tree}
tree.optim
```

Bootstrap on optimised model. The program resampled the nucleotides in
the alignment (columns in the input matrix).

```{r bootstrap, echo=FALSE, results='hide', message=FALSE, warning=FALSE}
bootstrap <- bootstrap.pml(tree.optim, optNni = TRUE)
```

```{r plot_bootstrap, fig.width=10, fig.height=5}
plotBS(tree.optim$tree, bootstrap, type = "phylogram")
```

`Mme_S` is more likely to affect patient `Mme_L`. The bootstrap support
is 98% (in my run with seed `42`), suggesting `Mme_S` and `Mme_L` share a more recent
ancestor (closer relationship) than `Mr D`
